logfire/_internal/integrations/system_metrics.py

```diff
@@ -0,0 +1,189 @@
+    if base == 'basic':
+        return BASIC_CONFIG
+    elif base == 'full':
+        return FULL_CONFIG
+    elif base is None:
+        return {}
+    else:
+        raise ValueError(f'Invalid base: {base}')
+
+
+def instrument_system_metrics(logfire_instance: Logfire, config: Config | None = None, base: Base = 'basic'):
+    config = {**get_base_config(base), **(config or {})}
+    SystemMetricsInstrumentor(config=config).instrument()  # type: ignore
+
+    if 'system.cpu.simple_utilization' in config:
+        measure_simple_cpu_utilization(logfire_instance)
+
+
+def measure_simple_cpu_utilization(logfire_instance: Logfire):
+    # The values of `process.cpu_percent()` are relative to the last time it was called.
+    # In particular, the first call will always return 0.
+    # So we need to call it once before we start measuring the actual utilization,
+    # and we need to use the same process object for all calls.
+    # TODO a similar problem exists with `psutil.cpu_percent()`
+    #   and thus for the OTEL instrumentation as well, and there it's harder to fix because
+    #   the previous values are separated by thread.
+    #   See https://github.com/open-telemetry/opentelemetry-python-contrib/issues/2797
+    process = psutil.Process()
+    process.cpu_percent()
+
+    def callback(_options: CallbackOptions) -> Iterable[Observation]:
+        percents: list[float] = [
+            # Average CPU usage across all cores.
+            # A high value is notable regardless of which core(s) this process is using.
+            psutil.cpu_percent(),
+            # CPU usage of this particular process.
+            # Can be greater than 100% if the process is using multiple cores.
+            # Will be less than 100% if multiple processes are using the same core,
+            # even if this process is using it at full capacity.
+            process.cpu_percent(),
+        ]
+        # CPU usage of the core this process is using, if available.
+        # This will be higher than `process.cpu_percent()` if multiple processes are using the same core.
+        # This requires `process.cpu_num()` which is only available on Linux,
+        # so we need to suppress the exception on other platforms.
+        with contextlib.suppress(Exception):
+            # Whether `Process.cpu_num` exists depends on the platform, and this affects pyright.
+            # So we can't use `# type: ignore` here, because on Linux it's not needed.
+            if not TYPE_CHECKING:  # pragma: no branch
+                cpu_num = process.cpu_num()
+                # `cpu_num` can be -1 on some platforms according to psutil.
+                if cpu_num >= 0:  # pragma: no branch
+                    percents.append(psutil.cpu_percent(percpu=True)[cpu_num])
+
+        # Return the highest of the three values.
+        # This means interpreting the value is not straightforward,
+        # but any unusual activity will show a notable spike, regardless of the infra setup.
```

Comment:
@samuelcolvin @dmontagu I'm having second thoughts about this. Maybe I should split this into 3 metrics:

- `system.cpu.simple_utilization`: just the average across all cores
- `process.runtime.cpu.utilization`: just the usage of this process - this is already an existing standard metric, it's part of the full config
- `process.core.cpu.utilization` (made up naming): just the usage of the core this process is using, if available.

Then the dashboard query would take the max of the three values at each recorded timestamp.

Pro: more detailed and interpretable data, less opinionated

Cons:

- The meaning of the defaults that the user must copy from the docs if they want to export a subset of those metrics goes from 3 metrics to 5:

```diff
logfire.instrument_system_metrics({
    'system.cpu.simple_utilization': None,
+  'process.runtime.cpu.utilization': None,
+  'process.core.cpu.utilization': None,
    'system.memory.utilization': ['available'],
    'system.swap.utilization': ['used'],
})
```

- More complex queries, particularly if users want to analyze CPU usage themselves.
- More data points. Probably not worth worrying about since it's still just 5 series, but it's 66% more than before.

Reply:
My weak preference would be to record:
* `system.cpu.simple_utilization`
*  and `process.runtime.cpu.utilization`

I think those are the things people will want in almost all cases.

Reply:
The problem with `process.runtime.cpu.utilization` is that if you have 5 processes running at full throttle on a single CPU, each will record 20% usage.

I suppose in that case `system.cpu.simple_utilization` will probably be high, I'm just not sure how reliable that is. I think some have mentioned that in some containerized environments on shared instances the CPU usage stats will include the usage of cores that the user doesn't actually have permission to use.

Reply:
For what it's worth, I think we should be summing CPU usage in most scenarios rather than averaging, and rely on the user being aware of the capacity of the machine.

Reply:
OK, I'm going to split it up into `system.cpu.simple_utilization` and `process.runtime.cpu.utilization` and drop the third value. However, `process.runtime.cpu.utilization` in the config doesn't actually mean that OTEL will export a metric called that, it'll name the metric something like `process.runtime.cpython.cpu.utilization` and it depends on the python implementation. The reason for this is in https://opentelemetry.io/docs/specs/semconv/general/metrics/#general-metric-semantic-conventions:

> Semantic ambiguity SHOULD be avoided. Use prefixed metric names in cases where similar metrics have significantly different implementations across the breadth of all existing metrics. For example, every garbage collected runtime has slightly different strategies and measures. Using a single set of metric names for GC, not divided by the runtime, could create dissimilar comparisons and confusion for end users. (For example, prefer `process.runtime.java.gc*` over `process.runtime.gc.*`.) Measures of many operating system metrics are similarly ambiguous.

I think OTEL made a bit of a mistake here. The way CPU usage is measured depends on the language, and garbage collection depends on the Python implementation, but all Python implementations measure CPU usage the same way via psutil. I'm thinking I should add a metric called `process.runtime.python.cpu.utilization` (that's `python`, not `cpython`) so that the name is stable. Then the dashboard query can just check `metric_name = 'process.runtime.python.cpu.utilization'`. I worry that something like `metric_name like 'process.runtime.%.cpu.utilization'` might not be reliably efficient, especially with variadic partitioning on metric name in FF.